File: backend/src/infrastructure/provider/ingestion/exporters/unreal/combat_designer_exporter.py
"""
combat_designer_exporter.py
Lightweight exporter script for Unreal Engine 5.x.
Runs locally in Unreal Editor Python environment to discover combat DataAssets,
extract raw combat parameters, compute sha256 checksums, and package into ExportBundle.
"""

import hashlib
import json
from datetime import datetime, timezone
from typing import Dict, Any, Union


EXPORTER_VERSION = "1.0.0"
PARSER_VERSION = "1.0.0"
FORMAT = "unreal-json"


def compute_sha256(content: Union[str, bytes]) -> str:
    if isinstance(content, str):
        content = content.encode("utf-8")
    return hashlib.sha256(content).hexdigest()


def compute_canonical_bundle_hash(checksums: Dict[str, str]) -> str:
    sorted_keys = sorted(checksums.keys())
    canonical_lines = [f"{k}:{checksums[k]}" for k in sorted_keys]
    canonical_str = "\n".join(canonical_lines)
    return compute_sha256(canonical_str)


def build_manifest(
    project_id: str,
    project_revision: str,
    workspace_id: str,
    file_contents: Dict[str, Union[str, bytes]],
    engine_version: str = "5.4"
) -> Dict[str, Any]:
    checksums = {}
    for path, content in file_contents.items():
        checksums[path] = compute_sha256(content)

    bundle_hash = compute_canonical_bundle_hash(checksums)

    manifest = {
        "schema_version": "1.0.0",
        "engine": "unreal",
        "engine_version": engine_version,
        "project_id": project_id,
        "project_revision": project_revision,
        "exporter_version": EXPORTER_VERSION,
        "parser_version": PARSER_VERSION,
        "format": FORMAT,
        "workspace_id": workspace_id,
        "exported_at": datetime.now(timezone.utc).isoformat(),
        "asset_count": len(file_contents),
        "checksums": checksums,
        "bundle_hash": bundle_hash,
    }
    return manifest

File: backend/src/infrastructure/provider/ingestion/exporters/unity/CombatDesignerExporter.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Generic;

#if UNITY_EDITOR
using UnityEditor;
using UnityEngine;
#endif

namespace CombatDesigner.Exporters.Unity
{
    /// <summary>
    /// Lightweight exporter script for Unity.
    /// Runs locally inside the user's Unity Editor to discover, checksum,
    /// and package combat ScriptableObjects into a standardized Export Bundle.
    /// </summary>
    public static class CombatDesignerExporter
    {
        public const string ExporterVersion = "1.0.0";
        public const string ParserVersion = "1.0.0";
        public const string Format = "unity-yaml-scriptable-object";

        [System.Serializable]
        public class ExportManifest
        {
            public string schema_version = "1.0.0";
            public string engine = "unity";
            public string engine_version = "2026.1";
            public string project_id = "default_project";
            public string project_revision = "git_head";
            public string exporter_version = ExporterVersion;
            public string parser_version = ParserVersion;
            public string format = Format;
            public string workspace_id = "default_workspace";
            public string exported_at;
            public int asset_count;
            public Dictionary<string, string> checksums = new Dictionary<string, string>();
            public string bundle_hash;
        }

        public static string ComputeSha256(byte[] bytes)
        {
            using (var sha = SHA256.Create())
            {
                byte[] hash = sha.ComputeHash(bytes);
                var sb = new StringBuilder();
                foreach (byte b in hash)
                {
                    sb.Append(b.ToString("x2"));
                }
                return sb.ToString();
            }
        }

        public static string ComputeCanonicalBundleHash(Dictionary<string, string> checksums)
        {
            var keys = new List<string>(checksums.Keys);
            keys.Sort(StringComparer.Ordinal);

            var sb = new StringBuilder();
            for (int i = 0; i < keys.Count; i++)
            {
                if (i > 0) sb.Append("\n");
                sb.Append(keys[i]).Append(":").Append(checksums[keys[i]]);
            }

            return ComputeSha256(Encoding.UTF8.GetBytes(sb.ToString()));
        }

        public static ExportManifest BuildManifest(
            string projectId,
            string projectRevision,
            string workspaceId,
            Dictionary<string, byte[]> fileContents)
        {
            var manifest = new ExportManifest
            {
                project_id = projectId,
                project_revision = projectRevision,
                workspace_id = workspaceId,
                exported_at = DateTime.UtcNow.ToString("o"),
                asset_count = fileContents.Count
            };

            foreach (var kvp in fileContents)
            {
                string hash = ComputeSha256(kvp.Value);
                manifest.checksums[kvp.Key] = hash;
            }

            manifest.bundle_hash = ComputeCanonicalBundleHash(manifest.checksums);
            return manifest;
        }
    }
}
